(DOCSP-28262): Kotlin: Add Manage Email/Password User page (#2762)
mongodb/docs-realm: source/sdk/kotlin/users/manage-email-password-users.txt

   :language: kotlin

.. _kotlin-confirm-a-new-users-email-address:

Confirm a New User's Email Address
----------------------------------

When you enable the :ref:`email/password provider <email-password-authentication>`,
in App Services, you select a confirmation method. The App Services email/password
confirmation service provides a token and token ID that you can get to the 
user through email or a custom Atlas Function. To confirm the user, 
provide that token and tokenId to `app.emailPasswordAuth.confirmUser()
<{+kotlin-sync-prefix+}io.realm.kotlin.mongodb.auth/-email-password-auth/confirm-user.html>`__:

.. literalinclude:: /examples/generated/kotlin/ManageEmailPasswordUsers.snippet.confirm-email-password-user.kt
   :language: kotlin

After you confirm the user, you can proceed to :ref:`login with email/password 
credentials <kotlin-email-password-login>`.

.. _kotlin-retry-user-confirmation:

Retry User Confirmation Methods
-------------------------------

The SDK provides methods to resend user confirmation emails or retry custom 
confirmation methods.

Resend a User Confirmation Email
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Resend a :ref:`confirmation email <auth-send-a-confirmation-email>`. Email/password 
> URLs expire after 30 minutes, so users who do not visit within that period 
> will need new emails to confirm their accounts.

Reviewer: ```suggestion
Resend a :ref:`confirmation email <auth-send-a-confirmation-email>`. The confirmation
tokens in each URL expire after 30 minutes. If a user does not follow the link and confirm
within that period, they will need to request a new confirmation email.
```

slight tweaks to be a bit more accurate that it's the tokens that expire

> To resend a user confirmation email, provide the email to `app.emailPasswordAuth.resendConfirmationEmail()

Reviewer: ```suggestion
To resend a user confirmation email, provide the user's email address to `app.emailPasswordAuth.resendConfirmationEmail()
```

<{+kotlin-sync-prefix+}io.realm.kotlin.mongodb.auth/-email-password-auth/resend-confirmation-email.html>`__:

.. literalinclude:: /examples/generated/kotlin/ManageEmailPasswordUsers.snippet.resend-password-confirmation-email.kt
   :language: kotlin

Retry a User Confirmation Function
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To retry a :ref:`custom user confirmation function <auth-run-a-confirmation-function>`,

> provide the email to `app.emailPasswordAuth.retryCustomConfirmation()

Reviewer: ```suggestion
provide the user's email address to `app.emailPasswordAuth.retryCustomConfirmation()
```

<{+kotlin-sync-prefix+}io.realm.kotlin.mongodb.auth/-email-password-auth/retry-custom-confirmation.html>`__:

.. literalinclude:: /examples/generated/kotlin/ManageEmailPasswordUsers.snippet.retry-custom-confirmation.kt
   :language: kotlin

.. _kotlin-reset-a-users-password:

Reset a User's Password
-----------------------

Resetting a user's password is a multi-step process. 

1. In your client app, you provide a UI for the user to reset their password. 
   Your App Services App can then send an email or run a custom function 
   to confirm the user's identity.
#. After confirming the user's identity, you can complete the password reset 
   request. 
#. After the password reset is complete, the user can log in using the new 
   password.

Select your preferred password reset method by going to:

1. Your Atlas App Services App
#. :guilabel:`Authentication`
#. :guilabel:`Authentication Providers`

> #. :guilabel:`Email/Password` - and press the :guilabel:`EDIT` button

Reviewer: I don't think we need to have BAAS configuration steps in the SDK docs. Could we just link to the [Password Resets](https://www.mongodb.com/docs/atlas/app-services/authentication/email-password/#password-resets) section instead?

Or, since we link to it below anyway, maybe we could just say that you need to configure it in the auth provider without written out steps.

Send a Password Reset Email
~~~~~~~~~~~~~~~~~~~~~~~~~~~

To send password reset emails to confirm the user's identity, 
you must configure your App to :ref:`send a password reset email 
<auth-send-a-password-reset-email>`.

To begin the password reset process, call `app.emailPasswordAuth.sendResetPasswordEmail() 
<{+kotlin-sync-prefix+}io.realm.kotlin.mongodb.auth/-email-password-auth/send-reset-password-email.html>`__ 
with the user's email. App Services sends an email to the user that contains 
a unique URL. The user must visit this URL within 30 minutes to confirm 
the reset.

.. literalinclude:: /examples/generated/kotlin/ManageEmailPasswordUsers.snippet.send-password-reset-email.kt
   :language: kotlin

After the user has visited the URL from the password reset email, call 

> call `app.emailPasswordAuth.resetPassword() 

Reviewer: ```suggestion
`app.emailPasswordAuth.resetPassword() 
```

call call

Author: Good catch - thanks!

When you configure your app to :ref:`run a password reset function 
<auth-run-a-password-reset-function>`, you define the function that 
should run when you call `app.emailPasswordAuth.callResetPasswordFunction() 
<{+kotlin-sync-prefix+}io.realm.kotlin.mongodb.auth/-email-password-auth/call-reset-password-function.html>`__
from the SDK. This function can take a username, a password, and any number 
of additional arguments. You can use these arguments to specify details 
like security question answers or other challenges that the user should 
pass to successfully complete a password reset.

You might prefer to use a custom password reset function when you want to
define your own password reset flows. For example, you might send a custom 
password reset email from a specific domain. Or you might use a service other 
than email to confirm the user's identity.

On the App Services side, you define the custom password reset function that
runs when you call this method. That function can return one of three possible
statuses:

- ``fail``
- ``pending``
- ``success``

A ``fail`` status is treated as an error by the SDK. The SDK ``callResetPasswordFunction()``

> does not take return values, so it does not return a ``pending`` or ``success``

Reviewer: "take return values" is weird phrasing to me. I think we want to say it returns `void` rather than indicating a status. We could also note that the function throws a `ServiceException` if the password reset fails.
